File: full-stack-lab/sdk/csharp/EchoClient.cs
using System.Diagnostics;
using System.IO;
using System.Net.Sockets;
using System.Text;
using OpenZiti;

namespace ZoSdk;

/// <summary>
/// Dials ZITI_SERVICE, sends "ping from cs\n", reads one line back, asserts it matches.
/// Prints exactly one RESULT line and exits 0 on success, non-zero on failure.
/// </summary>
public static class EchoClient
{
    public static int Run()
    {
        string identityFile = Env.IdentityFile;
        string serviceName  = Env.ServiceName;
        string target       = Env.Target(serviceName);

        var sw = Stopwatch.StartNew();
        try
        {
            var ctx    = Env.LoadContext(identityFile);
            var socket = new ZitiSocket(SocketType.Stream);
            API.Connect(socket, ctx, serviceName, "");

            using var ns = socket.ToNetworkStream();
            using var r  = new StreamReader(ns, Encoding.UTF8, leaveOpen: true);
            using var w  = new StreamWriter(ns, Encoding.UTF8, leaveOpen: true) { AutoFlush = true };

            const string ping = "ping from cs";
            w.WriteLine(ping);

            string? reply = r.ReadLine();
            sw.Stop();

            if (reply == ping)
            {
                Console.WriteLine($"RESULT ok echo cs->{target} {sw.ElapsedMilliseconds}ms");
                return 0;
            }

            Console.WriteLine($"RESULT fail echo cs->{target} unexpected-reply:{reply ?? "<null>"}");
            return 1;
        }
        catch (Exception ex)
        {
            sw.Stop();
            string reason = ex.Message.Replace(' ', '-');
            Console.WriteLine($"RESULT fail echo cs->{target} {reason}");
            return 2;
        }
    }
}

File: full-stack-lab/sdk/java/src/main/java/org/zo/interop/EchoClient.java
package org.zo.interop;

import org.openziti.ZitiConnection;
import org.openziti.ZitiContext;

import java.nio.charset.StandardCharsets;

/**
 * Echo client: dials ZITI_SERVICE, sends "ping from java\n", reads a line back, OK if it matches.
 */
final class EchoClient {

    static void run() throws Exception {
        String service = Env.service();
        String target = Env.target(service);
        long start = System.currentTimeMillis();

        try {
            ZitiContext ctx = Env.context();
            String msg = "ping from " + Env.LANG + "\n";

            try (ZitiConnection conn = ctx.dial(service)) {
                conn.write(msg.getBytes(StandardCharsets.UTF_8));

                String line = readLine(conn);
                long ms = System.currentTimeMillis() - start;

                if (msg.equals(line) || msg.stripTrailing().equals(line.stripTrailing())) {
                    Result.ok("echo", target, ms);
                } else {
                    Result.fail("echo", target, "mismatch: got '" + line.stripTrailing() + "'");
                }
            }
        } catch (Exception e) {
            Result.fail("echo", target, e.getClass().getSimpleName() + ": " + e.getMessage());
        }
    }

    private static String readLine(ZitiConnection conn) throws Exception {
        StringBuilder sb = new StringBuilder();
        byte[] one = new byte[1];
        while (true) {
            int n = conn.read(one, 0, 1);
            if (n < 0) {
                break;
            }
            if (n == 0) {
                continue;
            }
            char c = (char) (one[0] & 0xff);
            sb.append(c);
            if (c == '\n') {
                break;
            }
        }
        return sb.toString();
    }
}
